### DISCUSSION TITLE
New approach to proposing feature and enhancement

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8457] [Archive]
Today, new features and enhancements are proposed through the developer mailing list, for larger changes requiring a design proposal in community repository. This is not working very well as the developer mailing list can be hard to follow. Design proposals  in the community repository allows commenting while the PR is open, but is then harder to follow-up on after they have been merged.

As part of enabling GitHub Discussions it opens up a greater place to gather conversations around new proposals, and with features like participants, labelling and reactions it provides a better way to organise proposals.

The proposal here is that we use Ideas category in GitHub Discussions as the source/inventory as ideas. When making a new proposal it should be done by opening a new discussions there, and the developer mailing list should only serve as a notification. The developer mailing list can be used for lower level implementation discussions, but in general the discussion around a new feature should be happening in GitHub Discussions.

For smaller change proposals the complete proposal can be outlined in the discussion itself, while for larger proposals a design proposal should still be sent to the community repository, and linked to from the discussion. The design proposal should not have  a status attached to it, and should be merged relatively quickly (< 2 weeks). While the design proposal PR is open discussions can happen there, but the intent is to close it quickly and have most of the conversation happening through GitHub Discussions.

As a side-not it is worth mentioning that we plan to move away from Keycloak JIRA project to GitHub Issues. This will further emphasis the relevance of GitHub Discussions having a single tool to manage discussions, issues, as well as PRs.

**Comment by stianst** [comment]
Here's an example of a proposal following the new proposed approach:
https://github.com/keycloak/keycloak/discussions/8455

**Comment by thomasdarimont** [comment]
I think it makes sense to use github discussions here. This could work as a preliminary stage for a more strict feature description or specification later. However, for specs I prefer a shared google doc (or any other collaborative editor) that allows to comment something in-place.

Additionally I'd propose to assign a unique identifier to a feature / proposal. How about following the principles from the Python (PIP: Python Enhancement Proposal) and Java (JEP: JDK Enhancement Proposal) world and introducing a Keycloak Enhancement Proposal (KEP) to describe the new functionality mentioned above?
Perhaps with a template inspired by this: https://openjdk.java.net/jeps/2 ?

**Reply by stianst** [reply]
I'm not really keen on Google Docs as I think that mostly works well when you're collaborating with others within the same organization, and less so when collaborating in an open source community.

My intent was that the discussion would serve as a unique identifier (and permlink) for a feature/proposal.

**Reply by jonm-bb** [reply]
I guess that the permalink to the discussion is as good an ID as any, although you do lose it if you move off GitHub discussions for any reason.
For specs I think Google Docs works well for a smaller group, so no objections using a discussion (attached to a PR if needed) for a wider community of commenters and collaborators.

**Comment by stianst** [comment]
PR here: https://github.com/keycloak/keycloak/pull/8503

**Reply by pskopek** [reply]
It makes sense to me. All contributors are already at GitHub so no issues with access or have rights to comment/contribute.
